src/v/iceberg/schema_protobuf.cc

```diff
@@ -0,0 +1,185 @@
+          return d == &msg;
+      });
+}
+
+struct_outcome
+struct_from_protobuf(const pb::Descriptor& msg, type_stack& stack) {
+    if (is_recursive_type(msg, stack)) {
+        return schema_conversion_exception(fmt::format(
+          "Protocol buffer field {} not supported", msg.DebugString()));
+    }
+    if (stack.size() > max_recursion_depth) {
+        return schema_conversion_exception(fmt::format(
+          "Protocol buffer field {} not supported", msg.DebugString()));
+    }
+    stack.push_back(&msg);
+    struct_type struct_t;
+    struct_t.fields.reserve(msg.field_count());
+    for (int i = 0; i < msg.field_count(); ++i) {
+        auto field = msg.field(i);
+        auto res = from_protobuf(*field, field->is_repeated(), stack);
+        if (res.has_error()) {
+            return res.error();
+        }
+        struct_t.fields.push_back(std::move(res).value());
+    }
+    return std::move(struct_t);
+}
+
+field_outcome from_protobuf(
+  const pb::FieldDescriptor& fd, bool is_repeated, type_stack& stack) {
+    if (fd.is_map()) {
+        auto mt = fd.message_type();
+        auto key_field = mt->map_key();
+        auto value_field = mt->map_value();
+        auto key_nested_res = from_protobuf(*key_field, false, stack);
+        if (key_nested_res.has_error()) {
+            return key_nested_res;
+        }
+        auto value_field_nested_res = from_protobuf(*value_field, false, stack);
+        if (value_field_nested_res.has_error()) {
+            return value_field_nested_res;
+        }
+        auto field_type = map_type::create(
+          key_field->number(),
+          std::move(std::move(key_nested_res).assume_value()->type),
+          value_field->number(),
+          field_required(value_field->is_optional()),
+          std::move(std::move(value_field_nested_res).assume_value()->type));
+        return success(fd, std::move(field_type));
+    }
+    if (is_repeated) {
+        auto field_type_res = from_protobuf(fd, false, stack);
+        if (field_type_res.has_error()) {
+            return field_type_res;
+        }
+
+        auto type = list_type::create(
+          fd.number(),
+          field_required::no,
+          std::move(std::move(field_type_res).assume_value()->type));
+        return success(fd, std::move(type));
+    }
+
+    switch (fd.type()) {
+    case pb::FieldDescriptor::Type::TYPE_BOOL:
+        return success(fd, boolean_type{});
+    case pb::FieldDescriptor::TYPE_DOUBLE:
+        return success(fd, double_type{});
+    case pb::FieldDescriptor::TYPE_FLOAT:
+        return success(fd, float_type{});
+    case pb::FieldDescriptor::TYPE_INT64:
+    /**
+     * We support 32 bits unsigned integers by casting them to the iceberg long
+     * type, the long type can hold the max value of uint32_t without an
+     * overflow. This isn't true for 64 bit unsigned integers and this is why we
+     * return a validation error in this case.
```

Comment:
Maybe mention also that we do this because this what what offical iceberg Java code does with Parquet unsigned numbers?